In [0]:
import pandas as pd
import pandas
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error 
from sklearn.linear_model import LassoLarsCV
import matplotlib.pyplot as plt
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [0]:
csvfile = 'drive/My Drive/Colab Notebooks/finalmaster-ratios.csv'
alldata = pd.read_csv(csvfile)
alldata.head()

# Purchases  B01001001   B01001002  ...  B19001015  B19001016  B19001017
0           22     206252  469.226965  ...  36.440765  23.446284  21.197485
1            7      61399  486.538869  ...  33.000508  33.169741  24.792689
2            3      73170  489.859232  ...  20.125056  11.890525  16.537397
3           94     251724  505.585483  ...  43.731067  38.851729  40.427349
4            0      37382  495.586111  ...  36.118530  31.603714  19.648989

[5 rows x 190 columns]

In [0]:
allvariablenames = list(alldata.columns.values)
listofallpredictors= allvariablenames[9:]

In [0]:
predictors = alldata[listofallpredictors]  
target = alldata['# Purchases']   
predictors
target

0       22
1        7
2        3
3       94
4        0
      ... 
727      4
728      3
729      3
730      1
731    507
Name: # Purchases, Length: 732, dtype: int64

In [0]:
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123)    

In [0]:
model=LassoLarsCV(cv=10,precompute=False)
model.fit(predictors,target)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 8 iterations, i.e. alpha=8.526e-01, with an active set of 8 regressors, and the smallest cholesky pivot element being 8.816e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 12 iterations, i.e. alpha=4.263e-01, with an active set of 12 regressors, and the smallest cholesky pivot element being 2.220e-16. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 12 iterations, i.e. alpha=4.263e-01, with an active set of 12 regressors, and the smallest cholesky pivot element

LassoLarsCV(copy_X=True, cv=10, eps=2.220446049250313e-16, fit_intercept=True,
            max_iter=500, max_n_alphas=1000, n_jobs=None, normalize=True,
            positive=False, precompute=False, verbose=False)

In [0]:
predictors_model=pd.DataFrame(listofallpredictors)
predictors_model.columns = ['label']
predictors_model['coeff'] = model.coef_

for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)

['B01001014' 0.6626591370361307]
['B01001036' 1.7477634015843677]
['B01001037' 2.2471543892392045]
['B01001038' 2.032958095569556]
['B02001005' 1.0444093395536234]
['B02001007' 0.0026501086395314334]
['B13014026' 0.6378886301460416]
['B13014027' 1.373831154949208]
['B13016008' 0.7544555704553156]
['B15002027' 4.107586791367716]
['B19001017' 1.8782666975455609]


Question 1. The code above is running a four loop. First we define some variables. The first line of code is defining a variable predictors_model that creates a data frame of the predictor variables and coeficents. The second line is putting the predictors as in under'label'. The next line is putting coefs from the model in predictors_model['coeff'] column. Next is the for loop. The first line is saying for every row in predictor_model iterate through each row and do the following; if the coeff of that row is greater than 0 print the row and the value. Its printing the label and coefs that have a coef greater than 0. 

Question 2. In areas where there are more females over 25 with no diploma we sell more bobo bars. In areas with females aged 35 to 39 we sell more bobo bars. In aeas with females aged 40 to 44 we sell more. Areas with households that make $200000 or more are selling mroe bobo bars.Areas with women 30 to 34 sell more bobo bars. Areas with women 15 to 50 years old, who did not have a brith in the past 12 months, and never married or is widowed or divorced and has a graduate or professional degree we sell more bobo bars. Areas with more asians sell more bobo bars. Areas with males 40 to 44 sell more bobo bars. Areas with women who had a child in the last year and are 40 to 44 sell more bobos. Areas with women aged 15 to 50 who has a birth in the last year with a bachelors degree sell more bobo bars. Areas with people reporting other race sell more bobos. 

Questin 3. The top two best predictor variables for sales are women 25 year and over -12th grade, no diploma. The second varibles is females aged 35 to 39.

In [0]:
train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
20955.59181858268


In [0]:
test_error = mean_squared_error(tar_test, model.predict(pred_test))
print ('test data MSE')
print(test_error)

test data MSE
36219.3364898267


The training and test set mean standard errors are different, test data is higher than training data error. This means the test data seems to be further off from the actual data becasue it has a larger MSE. The training data is still off from the actual data but not as bad as the test data. This could indicate there is more vairability in the test data. 

In [0]:
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.27693842762902243


question 5. We can say that census data overall can help predict 27.7% of sales. The r- squared value measures the propotion of variance for a dependent variable explained by an independent variable. 

In [0]:
print("y interecept:")
print(model.intercept_)

y interecept:
30.487098362376265


the y intercept represents the sales when the predictors are 0. in this cases it would be the sales given no predictor variables or no information on the demogrphics or census.  